common/app/navigation/Navigation.scala

```diff
@@ -0,0 +1,157 @@
+package navigation
+
+import common.{Edition, Navigation, editions}
+import _root_.model.Page
+import NavLinks._
+
+import scala.annotation.tailrec
+
+case class Subnav(parent: Option[NavLink], children: Option[Seq[NavLink]], hasSecondaryParent: Boolean = false, hasTertiary: Boolean = false)
+
+sealed trait NavNode {
+  def children: Seq[NavLink]
+}
+
+case class NavLink(
+  title: String,
+  url: String,
+  longTitle: String = "",
+  // TODO: Shouldn't need iconName. Remove, and just make the first NavLink on mobile
+  iconName: String = "",
+  children: Seq[NavLink] = Nil
+) extends NavNode
+
+
+case class NavRoot private(val children: Seq[NavLink]) extends NavNode {
+
+  def getOtherPillarsFromEdition(edition: Edition): Seq[NavLink] = {
+    Edition.others(edition).flatMap( edition => {
+      NavRoot(edition).children
+    })
+  }
+
+  def findDescendantByUrl(url: String, edition: Edition): Option[NavLink] = {
+
+    @tailrec
+    def find(children: Seq[NavLink]): Option[NavLink] = {
+      children match {
+        case Nil => None
+        case head :: tail if (head.url == url) => Some(head)
+        case head :: tail => find(tail ++ head.children)
+      }
+    }
+    // If the link isn't found within the current edition, check other editions
+    find(children).orElse(find(getOtherPillarsFromEdition(edition)))
+  }
+
+  def findParentByCurrentNavLink(currentNavLink: NavLink, edition: Edition): Option[NavLink] = {
+    // Football is currently in the News Pillar and the Sport pillar, however we don't want the parent to be News.
+    def footballDoesNotBelongInNews(parentTitle: String): Boolean = {
+      currentNavLink.title == "football" && parentTitle == "news"
+    }
+
+    @tailrec
+    def find(children: Seq[NavLink]): Option[NavLink] = {
+      children match {
+        case Nil => find(getOtherPillarsFromEdition(edition))
+        case head :: tail if (head == currentNavLink) => Some(head) // If primary link, use that
+        case head :: tail if (footballDoesNotBelongInNews(head.title)) => find(tail ++ head.children)
+        case head :: tail if head.children.contains(currentNavLink) => Some(head)
+        case head :: tail => find(tail ++ head.children)
+      }
+    }
+    // If the parent isn't found within the current edition, check other editions
+    find(children).orElse(find(getOtherPillarsFromEdition(edition)))
+  }
+
+  def getPillar(pillars: Seq[NavLink], currentParent: NavLink, edition: Edition): NavLink = {
+    if(pillars.contains(currentParent)) {
+      currentParent
+    } else {
+      // TODO: should we have default pillar that is not UK (like a blank one)?
+      findParentByCurrentNavLink(currentParent, edition).getOrElse(ukNewsPillar)
+    }
+  }
+
+  def getSubnav(currentNavLink: Option[NavLink], currentParent: NavLink, currentPillar: NavLink): Subnav = {
```

Comment:
`currentParent` and `currentPillar` can be deducted based on the current `NavLink`. Why do we have to pass them as param?